services/worker-manager/src/providers/aws.js

```diff
@@ -0,0 +1,306 @@
+    providerId,
+    monitor,
+    rootUrl,
+    Worker,
+    WorkerPool,
+    WorkerPoolError,
+    estimator,
+    validator,
+    notify,
+    providerConfig,
+  }) {
+    super({
+      providerId,
+      monitor,
+      rootUrl,
+      Worker,
+      WorkerPool,
+      WorkerPoolError,
+      estimator,
+      validator,
+      notify,
+    });
+    this.configSchema = 'config-aws';
+    this.ec2iid_RSA_key = fs.readFileSync(path.resolve(__dirname, 'aws-keys/RSA-key-forSignature')).toString();
+    this.providerConfig = providerConfig;
+  }
+
+  /*
+   This method is used to setup permissions for the EC2 instances
+   */
+  async setup() {
+  }
+
+  async provision({workerPool}) {
+    const {workerPoolId} = workerPool;
+
+    if (!workerPool.providerData[this.providerId] || workerPool.providerData[this.providerId].running === undefined) {
+      await workerPool.modify(wt => {
+        wt.providerData[this.providerId] = wt.providerData[this.providerId] || {};
+        wt.providerData[this.providerId].running = wt.providerData[this.providerId].running || 0;
+      });
+    }
+
+    const config = this.chooseConfig({possibleConfigs: workerPool.config.launchConfigs});
+
+    aws.config.update({region: config.region});
+    aws.config.logger = console;
+    const ec2 = new aws.EC2({
+      apiVersion: AWS_API_VERSION,
+      credentials: this.providerConfig.credentials,
+    });
+
+    const toSpawn = await this.estimator.simple({
+      workerPoolId,
+      minCapacity: config.launchConfig.MinCount,
+      maxCapacity: config.launchConfig.MaxCount,
+      capacityPerInstance: config.capacityPerInstance,
+      running: workerPool.providerData[this.providerId].running,
+    });
+
+    const userData = {
+      rootUrl: this.rootUrl,
+      workerPoolId,
+      providerId: this.providerId,
+      workerGroup: this.providerId,
+    };
+
+    let spawned;
+    try {
+      spawned = await ec2.runInstances({
+        ...config.launchConfig,
+
+        // please make sure this string is no more than 1024 chars long
+        // https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html#instancedata-add-user-data
+        UserData: JSON.stringify(userData).toString('base64'),
+
+        MaxCount: config.launchConfig.MaxCount ? Math.max(toSpawn, config.launchConfig.MaxCount) : toSpawn,
+        MinCount: config.launchConfig.MinCount ? Math.min(toSpawn, config.launchConfig.MinCount) : toSpawn,
+        TagSpecifications: [
+          ...(config.launchConfig.TagSpecifications ? config.launchConfig.TagSpecifications : []),
+          {
+            ResourceType: 'instance',
+            Tags: [
+              {
+                Key: 'Provider',
+                Value: `wm-${this.providerId}`,
+              }, {
+                Key: 'Owner',
+                Value: workerPool.owner,
+              }],
+          },
+        ],
+      }).promise();
+    } catch (e) {
+      this.monitor.err(`🚨 Error calling AWS API: ${e}`);
```

Comment:
Let's leave emoji out of the official error messages :)

Reply:
lol ok